File: src/main/java/net/volcano/jdacommands/interfaces/PermissionProvider.java
package net.volcano.jdacommands.interfaces;

import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.User;
import net.volcano.jdacommands.model.command.CommandEvent;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Set;

public interface PermissionProvider {
	
	Set<String> getPermissions(String userId);
	
	Set<String> getPermissions(String userId, @Nullable String guildId);
	
	default Set<String> getPermissions(User user, @Nullable Guild guild) {
		return getPermissions(user.getId(), guild == null ? null : guild.getId());
	}
	
	default Set<String> getPermissions(CommandEvent event) {
		return getPermissions(event.getAuthor(), event.isFromGuild() ? event.getGuild() : null);
	}
	
	boolean hasPermissions(Collection<String> permissions, String userId, @Nullable String guildId);
	
	default boolean hasPermissions(Collection<String> permissions, User user, @Nullable Guild guild) {
		return hasPermissions(permissions, user.getId(), guild == null ? null : guild.getId());
	}
	
}

File: src/main/java/net/volcano/jdacommands/interfaces/PermissionProvider.kt
package net.volcano.jdacommands.interfaces

import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.volcano.jdacommands.permissions.PermissionHolder

interface PermissionProvider {

	fun getPermissions(event: MessageReceivedEvent): Set<PermissionHolder> {
		return getPermissions(
			event.author,
			if (event.isFromGuild) event.guild else null,
			if (event.isFromGuild) event.textChannel else null
		)
	}

	fun getPermissions(user: User, guild: Guild? = null, channel: TextChannel? = null): Set<PermissionHolder> {
		return getPermissions(user.id, guild?.id, channel?.id)
	}

	fun getPermissions(userId: String, guildId: String? = null, channelId: String? = null): Set<PermissionHolder>

	fun isOverriding(userId: String): Boolean

	fun startOverriding(userId: String)

	fun stopOverriding(userId: String)

}